firebase-perf/src/test/java/com/google/firebase/perf/application/FragmentStateMonitorTest.java

```diff
@@ -88,79 +88,75 @@ public void setUp() {
     doAnswer((Answer<Timer>) invocationOnMock -> new Timer(currentTime)).when(clock).getTime();
 
     DeviceCacheManager.clearInstance();
-    ConfigResolver.clearInstance();
+    doReturn(true).when(configResolver).isPerformanceMonitoringEnabled();
+
+    doNothing().when(recorder).start();
+    doNothing().when(recorder).startFragment(any());
 
-    ConfigResolver configResolver = ConfigResolver.getInstance();
-    configResolver.setDeviceCacheManager(new DeviceCacheManager(new FakeDirectExecutorService()));
-    ConfigResolver spyConfigResolver = spy(configResolver);
-    doReturn(true).when(spyConfigResolver).isPerformanceMonitoringEnabled();
-    this.configResolver = spyConfigResolver;
+    // doReturn(Optional.of(frameCounts1)).when(recorder).stopFragment(any());
+    // doReturn(Optional.of(frameCounts2)).when(recorder).stopFragment(any());
+
+    activity = createFakeActivity(true);
 
     // fmaMetrics1 should have 1+3+1=5 total frames, 3+1=4 slow frames, and 1 frozen frames.
-    SparseIntArray sparseIntArray = new SparseIntArray();
-    sparseIntArray.append(1, 1);
-    sparseIntArray.append(17, 3);
-    sparseIntArray.append(800, 1);
-    fmaMetrics1[FrameMetricsAggregator.TOTAL_INDEX] = sparseIntArray;
+    frameCounts1 = new PerfFrameMetrics(9, 5, 3);
 
     // fmaMetrics2 should have 5+5+4=14 total frames, 5+4=9 slow frames, and 4 frozen frames.
-    sparseIntArray = new SparseIntArray();
-    sparseIntArray.append(1, 5);
-    sparseIntArray.append(18, 5);
-    sparseIntArray.append(800, 4);
-    fmaMetrics2[FrameMetricsAggregator.TOTAL_INDEX] = sparseIntArray;
+    frameCounts2 = new PerfFrameMetrics(14, 9, 4);
   }
 
   /************ Trace Creation Tests ****************/
 
+  @Test
+  public void fragmentTraceName_validFragment_validFragmentScreenTraceNameGenerated() {
+    FragmentStateMonitor monitor =
+        new FragmentStateMonitor(clock, mockTransportManager, appStateMonitor, recorder);
+    Fragment testFragment = new FragmentMonitorMockFragment();
+    Assert.assertEquals(
+        "_st_FragmentMonitorMockFragment", monitor.getFragmentScreenTraceName(testFragment));
+  }
+
   @Test
   public void lifecycleCallbacks_differentFrameMetricsCapturedByFma_logFragmentScreenTrace() {
     FragmentStateMonitor monitor =
-        new FragmentStateMonitor(clock, mockTransportManager, appStateMonitor, fma);
-    when(fma.getMetrics()).thenReturn(fmaMetrics1);
+        new FragmentStateMonitor(clock, mockTransportManager, appStateMonitor, recorder);
+    doReturn(Optional.of(frameCounts1)).when(recorder).stopFragment(any());
```

Comment:
Is there a specific reasons that you want to use `doReturn..when` instead of `when...thenReturn`? 
go/java-testing/mocking#stubbing-with-when-does-not-always-work suggests "Tip: Use Mockito.doReturn(value).when(mockObject).method(arguments) in cases where when(...) does not work. Avoid it otherwise (style rule, it is harder to read)."
Applies to other places in this test file.

Reply:
Good reference point. Thanks Jeremy!

Reply:
Something I learnt. Mockito.doReturn when used on spied objects, does not make the function call at all. Whereas, Mockito.when will call the function and if there are exceptions, we need to handle that. So had to revert back this change in one of the places.

Reply:
A good thread on that is here - https://stackoverflow.com/questions/20353846/mockito-difference-between-doreturn-and-when